### ISSUE TITLE
Apply symbols visibility settings

### ISSUE THREAD

**Issue by Piccions** [ROOT-Sim/core#33]
Right now, ROOT-Sim libraries expose all their internal global symbols to library users.
We should change the build process to that global symbols are hidden to users and annotate the actual API functions as externally visible (since our API symbols start with a capital letter, maybe this could be done automatically by a script).
This problem pertains not only shared libraries symbol visibility but also LTO visibility (see https://clang.llvm.org/docs/LTOVisibility.html)

**Comment by alessandropellegrini** [comment]
I think that setting `-fvisibility=hidden` as a global compilation flag should be sufficient also for LTO visibility, in our case.

I wouldn't go with the script, rather with something like:
```
#define public __attribute__((visibility("default")))
```
and then manually annotate the API. I think this makes the code base much more readable and clear, rather than forcing people to discover that there is some obscure script in the workflow that does some visibility magic. We could discuss whether `public` is a good name though (been in the Java world too much for a while now :smile: ).

**Comment by alessandropellegrini** [comment]
Shall we close this one?

**Comment by Piccions** [comment]
It depends: do we plan to support the build of a dynamic library? If so, we still didn't verify that symbol visibility settings are correct. IIRC we didn't annotate any API with the default visibility attribute.

**Comment by alessandropellegrini** [comment]
I agree on both points. The benefits of having a dynamic library are still unclear currently. The annotations give an advantage only in that case. Given that, I don't know if having a stale issue open for years makes sense, hence my idea to close it. But I have nothing against it. If we keep it open, at least we should remove it from release 3.0.0.
